File: clients/fsharp/Model/BuildingProperties.fs
#nowarn "0058"

namespace SpbAiChamp
namespace SpbAiChamp.Model

open SpbAiChamp

/// Building properties
type BuildingProperties = {
    /// Resources required for building
    BuildResources: Map<Model.Resource, int>;
    /// Max health points of the building
    MaxHealth: int;
    /// Max number of workers in the building
    MaxWorkers: int;
    /// Resources required to start another task
    WorkResources: Map<Model.Resource, int>;
    /// Whether performing a task spawn new workers
    ProduceWorker: bool;
    /// Resource produced when performing a task
    ProduceResource: option<Model.Resource>;
    /// Amount of resources/workers produced when performing one task
    ProduceAmount: int;
    /// Score points given for performing one task
    ProduceScore: int;
    /// Whether building is harvesting. In this case resource can only be produced if it is harvestable on the planet
    Harvest: bool;
    /// Amount of work needed to finish one task
    WorkAmount: int;
} with

    /// Write BuildingProperties to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.BuildResources.Count
        this.BuildResources |> Map.iter (fun key value ->
            writer.Write (int key)
            writer.Write value )
        writer.Write this.MaxHealth
        writer.Write this.MaxWorkers
        writer.Write this.WorkResources.Count
        this.WorkResources |> Map.iter (fun key value ->
            writer.Write (int key)
            writer.Write value )
        writer.Write this.ProduceWorker
        match this.ProduceResource with
            | Some value ->
                writer.Write true
                writer.Write (int value)
            | None -> writer.Write false
        writer.Write this.ProduceAmount
        writer.Write this.ProduceScore
        writer.Write this.Harvest
        writer.Write this.WorkAmount
        ()

    /// Read BuildingProperties from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        BuildResources = [for _ in 1 .. reader.ReadInt32() do
                             let key = reader.ReadInt32() |> enum
                             let value = reader.ReadInt32()
                             yield (key, value) ] |> Map.ofList
        MaxHealth = reader.ReadInt32()
        MaxWorkers = reader.ReadInt32()
        WorkResources = [for _ in 1 .. reader.ReadInt32() do
                            let key = reader.ReadInt32() |> enum
                            let value = reader.ReadInt32()
                            yield (key, value) ] |> Map.ofList
        ProduceWorker = reader.ReadBoolean()
        ProduceResource = match reader.ReadBoolean() with
                              | true -> Some(reader.ReadInt32() |> enum)
                              | false -> None
        ProduceAmount = reader.ReadInt32()
        ProduceScore = reader.ReadInt32()
        Harvest = reader.ReadBoolean()
        WorkAmount = reader.ReadInt32()
    }

File: clients/rust/src/model/building_properties.rs
use super::*;

/// Building properties
#[derive(Clone, Debug)]
pub struct BuildingProperties {
    /// Building type that this building can be upgraded from
    pub base_building: Option<model::BuildingType>,
    /// Resources required for building
    pub build_resources: std::collections::HashMap<model::Resource, i32>,
    /// Max health points of the building
    pub max_health: i32,
    /// Max number of workers in the building
    pub max_workers: i32,
    /// Resources required to start another task
    pub work_resources: std::collections::HashMap<model::Resource, i32>,
    /// Whether performing a task spawn new workers
    pub produce_worker: bool,
    /// Resource produced when performing a task
    pub produce_resource: Option<model::Resource>,
    /// Amount of resources/workers produced when performing one task
    pub produce_amount: i32,
    /// Score points given for performing one task
    pub produce_score: i32,
    /// Whether building is harvesting. In this case resource can only be produced if it is harvestable on the planet
    pub harvest: bool,
    /// Amount of work needed to finish one task
    pub work_amount: i32,
}

impl trans::Trans for BuildingProperties {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.base_building.write_to(writer)?;
        self.build_resources.write_to(writer)?;
        self.max_health.write_to(writer)?;
        self.max_workers.write_to(writer)?;
        self.work_resources.write_to(writer)?;
        self.produce_worker.write_to(writer)?;
        self.produce_resource.write_to(writer)?;
        self.produce_amount.write_to(writer)?;
        self.produce_score.write_to(writer)?;
        self.harvest.write_to(writer)?;
        self.work_amount.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let base_building: Option<model::BuildingType> = trans::Trans::read_from(reader)?;
        let build_resources: std::collections::HashMap<model::Resource, i32> = trans::Trans::read_from(reader)?;
        let max_health: i32 = trans::Trans::read_from(reader)?;
        let max_workers: i32 = trans::Trans::read_from(reader)?;
        let work_resources: std::collections::HashMap<model::Resource, i32> = trans::Trans::read_from(reader)?;
        let produce_worker: bool = trans::Trans::read_from(reader)?;
        let produce_resource: Option<model::Resource> = trans::Trans::read_from(reader)?;
        let produce_amount: i32 = trans::Trans::read_from(reader)?;
        let produce_score: i32 = trans::Trans::read_from(reader)?;
        let harvest: bool = trans::Trans::read_from(reader)?;
        let work_amount: i32 = trans::Trans::read_from(reader)?;
        Ok(Self {
            base_building,
            build_resources,
            max_health,
            max_workers,
            work_resources,
            produce_worker,
            produce_resource,
            produce_amount,
            produce_score,
            harvest,
            work_amount,
        })
    }
}